docs/codeql/codeql-language-guides/customizing-library-models-for-cpp.rst

```diff
@@ -0,0 +1,184 @@
+- ``sinkModel(namespace, type, subtypes, name, signature, ext, input, kind, provenance)``. This is used to model sinks where tainted data may be used in a way that makes the code vulnerable.
+- ``summaryModel(namespace, type, subtypes, name, signature, ext, input, output, kind, provenance)``. This is used to model flow through elements.
+
+The extensible predicates are populated using the models defined in data extension files.
+
+Example of custom model definitions
+------------------------------------
+
+The examples in this section are taken from the standard CodeQL CPP query pack published by GitHub. They demonstrate how to add tuples to extend extensible predicates that are used by the standard queries.
+
+Example: Taint source from the ``boost::asio`` namespace
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example shows how the CPP query pack models the return value from the ``read_until`` function as a ``remote`` source.
+
+.. code-block:: cpp
+
+   boost::asio::read_until(socket, recv_buffer, '\0', error);
+
+We need to add a tuple to the ``sourceModel``\(namespace, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/cpp-all
+         extensible: sourceModel
+       data:
+         - ["boost::asio", "", False, "read_until", "", "", "Argument[*1]", "remote", "manual"]
+
+Since we are adding a new source, we need to add a tuple to the ``sourceModel`` extensible predicate.
+The first five values identify the callable (in this case a free function) to be modeled as a source.
+
+- The first value ``"boost::asio"`` is the namespace name.
+- The second value ``""`` is the name of the type (class) that contains the method. Because we're modelling a free function, the type is left blank.
+- The third value ``False`` is a flag that indicates whether or not the sink also applies to all overrides of the method. For a free function, this should be ``False``.
+- The fourth value ``"read_until"`` is the function name.
+- The fifth value is the function input type signature, which can be used to narrow down between functions that have the same name. In this case, we want the model to include all functions in ``boost::asio`` called ``read_until``.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the source.
+
+- The seventh value ``"Argument[*1]"`` is the ``access path``, which means that the sink is the first indirection (or pointed-to value, ``*``) of the second argument (``Argument[1]``) passed to the function.
```

Comment:
Do other languages also call this "access path"? This is not really aligned with what an "access path" in the dataflow library is (or the way we talk about access paths internally). I'd much rather we called this "input specification" and "input specification" or something like that, but if other languages are already calling this "access paths" I guess we have to live with it 😅

Reply:
It was described as an "access path" in the csharp doc this is based on.  I'm happy to rephrase if "input specification" is preferred?

Reply:
I do prefer "input specification", but I suppose there's some benefits to language consistency 🤔 I'm happy with whatever you decide on this one

Reply:
Done.  Please could you check you're happy with the new phrasing?